ADTs, Control flow and Generics sections for preliminaries (#9)
FLAGlab/bounded-const-generics-rust: document/chapters/preliminaries.tex

In principle, every reference needs a lifetime annotation. However, this is avoided by a process known as lifetime elision, where lifetimes are inferred automatically by the compiler. Even then, in certain cases the programmer might need to add such annotations. However, this will be discussed in the generics section.

\section{Algebraic data types}

> Rust has both sum and product algebraic data types in the form of structures and enumerations respectively.

Reviewer: ```suggestion
Rust has both sum and product algebraic data types~\cite{REF} in the form of structures and enumerations respectively.
```

Author: We need a reference about the notion of algebraic data types? or about the fact that Rust has algebraic data types?

> Structures are named sum types, where an struct type has a fixed set of named fields. When declaring a new value of an struct type, all its field must be given as shown in TODO. Enumerations are named product types, where an enum type has a fixed set of named variants. When declaring a new value of an enum type, one single variant must be chosen as shown in TODO.

Reviewer: ```suggestion
Structures are named sum types, where an struct type has a fixed set of named fields. When declaring a new value of an struct type, all its fields must be given as shown in TODO. Enumerations are named product types, where an enum type has a fixed set of named variants. When declaring a new value of an enum type, one single variant must be chosen as shown in TODO.
```

> It is also possible to add associated functions to any type using the \texttt{impl} keyword in an object oriented programming style. Each associated function could or could not use an instance of a given type, just as instance and static methods in languages such as Java.

Reviewer: ```suggestion
It is also possible to add associated functions to any type using the \texttt{impl} keyword in an object oriented programming style. Each associated function could or could not use an instance of a given type. This is similar, for example, to the way instance and static methods are defined in Java.
```

\section{Control flow}

> Rust has both the usual \texttt{if/else} conditional and \texttt{while} loop statements as shown in TODO and TODO respectively. \texttt{for} loops are iterator based, where the variable to be iterated must implement the \texttt{Iterator} trait.

Reviewer: ```suggestion
Control flow in Rust is realized using the common \texttt{if/else} conditional and \texttt{while} loop statements, as in other programming languages. \texttt{for} loops are iterator based, where the variable to be iterated must implement the \texttt{Iterator} trait.
```

Rust has pattern matching capabilities thanks to its functional heritage, pattern matching in Rust can be used to match specific values or to destructure tuples, arrays, enumerations or structures. There are three statements to do pattern matching: \texttt{match}, \texttt{if let} and \texttt{while let}. Examples for these statemens can be found in TODO, TODO and TODO respectively. 
\section{Generics}

> From an external perspective, generic types in Rust are quite similar to the generic types in Java, i.e. every type can have generic type parameters in order to reduce code duplication and such parameters are specified between angled brackets. However, there are three main differences:

Reviewer: ```suggestion
From an external perspective, generic types in Rust are quite similar to the generic types in Java --that is, every type can have generic type parameters in order to reduce code duplication and such parameters are specified between angled brackets. However, there are three main differences between the two implementations:
```

\begin{itemize}
    \item Rust allows lifetimes as generic parameters. These are used when the borrow checker cannot infer a proper lifetime for a variable, and thus explicit lifetime annotations are required. An example of this can be found in TODO.
    \item In Java, generic parameters can be bounded by forcing them to be instances of a class or interface. In Rust, which is not an object oriented language, bounds are done over traits instead, as in TODO.

>     \item Internally, generics in Java are implemented doing type erasure, where each generic parameter bounds are checked, and then the Java compiler forgets about the type of such paremeter and uses dynamic dispatch over the type \texttt{Object}. On the other hand, Rust uses monomorphization, where the compiler builds a new type specialized for each instance of a generic parameter making all the dispatch completely static, as shown in TODO. \footnote{Rust allows type erasure via trait objects. However, this goes is beyond the scope of this work.}

Reviewer: ```suggestion
    \item Internally, generics in Java are implemented doing type erasure, where each generic parameter bounds are checked, and then the Java compiler forgets about the type of such paremeter and uses dynamic dispatch over the type \texttt{Object}. On the other hand, Rust uses monomorphization, where the compiler builds a new type specialized for each instance of a generic parameter making all the dispatch completely static, as shown in TODO.\footnote{Rust allows type erasure via trait objects. However, this goes is beyond the scope of this work.}
```
